frame/election-provider-multi-phase/src/signed.rs

```diff
@@ -0,0 +1,814 @@
+		Some(self.cmp(other))
+	}
+}
+
+pub type BalanceOf<T> =
+	<<T as Config>::Currency as Currency<<T as frame_system::Config>::AccountId>>::Balance;
+pub type PositiveImbalanceOf<T> = <<T as Config>::Currency as Currency<
+	<T as frame_system::Config>::AccountId,
+>>::PositiveImbalance;
+pub type NegativeImbalanceOf<T> = <<T as Config>::Currency as Currency<
+	<T as frame_system::Config>::AccountId,
+>>::NegativeImbalance;
+pub type SignedSubmissionOf<T> =
+	SignedSubmission<<T as frame_system::Config>::AccountId, BalanceOf<T>, CompactOf<T>>;
+
+pub type SubmissionIndicesOf<T> =
+	BoundedBTreeMap<ElectionScore, u32, <T as Config>::SignedMaxSubmissions>;
+
+/// Mask type which pretends to be a set of `SignedSubmissionOf<T>`, while in fact delegating to the
+/// actual implementations in `SignedSubmissionIndices<T>`, `SignedSubmissionsMap<T>`, and
+/// `SignedSubmissionNextIndex<T>`.
+#[cfg_attr(feature = "std", derive(DebugNoBound))]
+pub struct SignedSubmissions<T: Config> {
+	indices: SubmissionIndicesOf<T>,
+	next_idx: u32,
+	insertion_overlay: BTreeMap<u32, SignedSubmissionOf<T>>,
+	deletion_overlay: BTreeSet<u32>,
+}
+
+impl<T: Config> SignedSubmissions<T> {
+	/// Get the signed submissions from storage.
+	pub fn get() -> Self {
+		SignedSubmissions {
+			indices: SignedSubmissionIndices::<T>::get(),
+			next_idx: SignedSubmissionNextIndex::<T>::get(),
+			insertion_overlay: BTreeMap::new(),
+			deletion_overlay: BTreeSet::new(),
+		}
+	}
+
+	/// Put the signed submissions back into storage.
+	pub fn put(self) {
+		SignedSubmissionIndices::<T>::put(self.indices);
+		SignedSubmissionNextIndex::<T>::put(self.next_idx);
+		for key in self.deletion_overlay {
+			SignedSubmissionsMap::<T>::remove(key);
+		}
+		for (key, value) in self.insertion_overlay {
+			SignedSubmissionsMap::<T>::insert(key, value);
+		}
+	}
+
+	/// Get the submission at a particular index.
+	fn get_submission(&self, idx: u32) -> Option<SignedSubmissionOf<T>> {
+		self.insertion_overlay
+			.get(&idx)
+			.cloned()
+			.or_else(|| SignedSubmissionsMap::<T>::try_get(idx).ok())
+	}
+
+	/// Take the submission at a particular index.
+	fn take_submission(&mut self, idx: u32) -> Option<SignedSubmissionOf<T>> {
+		self.insertion_overlay.remove(&idx).or_else(|| {
+			if self.deletion_overlay.contains(&idx) {
+				None
+			} else {
+				self.deletion_overlay.insert(idx);
+				SignedSubmissionsMap::<T>::try_get(idx).ok()
+			}
+		})
+	}
+
+	/// Iterate through the set of signed submissions in order of increasing score.
+	pub fn iter(&self) -> impl '_ + Iterator<Item = SignedSubmissionOf<T>> {
+		self.indices.iter().map(move |(_score, idx)| {
+			self.get_submission(*idx).expect("SignedSubmissions must remain internally consistent")
+		})
+	}
+
+	/// Empty the set of signed submissions, returning an iterator of signed submissions in
+	/// arbitrary order
+	pub fn drain(&mut self) -> impl '_ + Iterator<Item = SignedSubmissionOf<T>> {
+		self.indices.clear();
```

Comment:
is this a mandatory usage of drain to drain the whole map ?

Currently storagemap::drain behvaior is: `drain().next().next()` remove only 2 entry from storage, no more.
So here if somebody calls `drain` but doesn't consume the whole iterator, then it is in a weird state.

Reply:
maybe the iterator should kill the map when dropped?

Reply:
Yes, `drain` is expected to drain the entire map. I think that having an explicit `struct Iter` which kills the map on `Drop` may be the right answer in the long run, but for now, I've just made the function private and verified that it's never terminated without consuming all elements.

Reply:
maybe the doc can say that all element must be consumed, otherwise the variable is in an invalid state, or the function could take self instead of &mut self

Reply:
it is a bit weird to me that all function are designed not touch the storage but this function do without saying so.

I see it is private, but it is accessible from this file, and we use it directly.

Also this function update the storage: SignedSubmissionNextIndex, SignedSubmissionsMap but not SignedSubmissionIndices
And update the fields: `indices` but not `next_index`.

Reply:
Yeah, agree. I've updated it again: 73071aaebba06036e4e566650e8a5ed47fc2fa97.